### ISSUE TITLE
Access control with ID token

### ISSUE THREAD

**Issue by salmanisd** [oauth2-proxy/oauth2-proxy#680]
I have been testing oauth2-proxy with nginx and using ory hydra as OIDC provider at the backend and everything is running very smoothly (thanks for that).

I have a question about implementing access control on specific nginx locations.
e.g) my current flow
1) try to access any resource on nginx
2) redirected by oauth2-proxy to my identity provider
3) enter your credentials, a userid token is issued by hydra
4) oauth2-proxy successfully authenticates and l login to my homepage
no issues till this point, everything is working fine

after logging in, suppose I see two apps. APP1 at location localhost/app1 and APP2 at location localhost/app2.
Maybe I only have permissions to view APP2 so when I click APP1 it should be denied.
Is it possible to implement something like this using oauth2-proxy?

e.g)adding the permissions(app1,app2) to the "claims" part of the id token which we got in step 4 and then use the same token to further refine permissions? (or maybe there's some other way)

**Comment by NickMeves** [comment]
At the moment there's nothing for claim based authorization rules in the OIDC provider.  This PR is taking a look at it (group claim specifically): https://github.com/oauth2-proxy/oauth2-proxy/pull/616

Other providers (Github, Gitlab, Google I think) have very rudimentary support for group based AuthZ -- but for the most part oauth2-proxy specializes in authentication and has delegated authorization to the backend apps (via passing details about the user in `X-Forwarded-*` user headers or IDToken via `Authorization` header.

I know @JoelSpeed is planning out much richer rules engine for authZ -- but I don't think we've gone down the path of path based authorization rules (in V6 now or planned for the future)

Only thing I can think of at the moment if you have multiple oauth2-proxies deployed for each app/path you want to authorize differently... But that sounds awful 🤦

**Comment by salmanisd** [comment]
I was thinking something around these lines(please correct me if I'm wrong)

1) authenticate with oauth2-proxy and get the idtoken
2) pass the access token from oauth2-proxy to nginx (using --pass-access-token)
3) in nginx for each app location e.g) localhost/app1 , localhost/app2 , replace oauth2-proxy's auth_request directive with own custom auth_request which takes the access token, sends it to /userinfo and then parses the response to see the permissions/claims.
(something like this https://www.nginx.com/blog/validating-oauth-2-0-access-tokens-nginx/)

Do you think this could without running into conflict with oauth2-proxy?

**Comment by salmanisd** [comment]
So I did some rough implementation and with some hardcoded values it seems to be working

     location / {
            auth_request /internal-auth/oauth2/auth;

            # Alternatively send the request to `start` to skip the provider button
            error_page 401 = http://oauth2-proxy.oauth2-proxy.localhost/oauth2/start?rd=$scheme://$host$request_uri;

            root /usr/share/nginx/html;
            index index.html index.htm;
      }

      location /app1 {
            #    auth_request /internal-auth/oauth2/auth;
            auth_request /gg;

            proxy_pass http://localhost:1890/app1;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
      }

      location = /gg {
            internal;
            js_content introspectAccessToken;
      }

      location = /_oauth2_send_request {
            internal;
            proxy_method POST;
            proxy_set_header content-type "application/json";
            proxy_set_header  Authorization "Bearer uaXjWIppT_Ybe5eLNlMOzM1LcrcKrysEljm7knksFcY.1rXTOGO-lEetOJEQnLV7qsFBC122yhUStZSbxen3OHw";
            proxy_pass        http://localhost:4444/userinfo;
      }
`
If I access localhost I have to login. 
If I access localhost/app1 then  access is granted or denied, depending on the value of the hardcoded access token above (Bearer ). Right now I just check the "sub" key in the response I get from /userinfo. For access control I will have to parse scope/claims out of the response.

The only issue right now is if I dont know how to take the access_token from oauth2-proxy and use it as the bearer token instead of the hardcoded one. 
My config file is 

```
http_address="0.0.0.0:4180"
cookie_secret="OQINaROshtE9TcZkNAm-5Zs2Pv3xaWytBmc5W7sPX7w="
provider="oidc"
email_domains="example.com"
oidc_issuer_url="http://localhost:4444/"
client_secret="b2F1dGgyLXByb3h5LWNsaWVudC1zZWNyZXQK"
client_id="oauth2-proxy"
cookie_secure="false"

set_xauthrequest="true"
upstreams="http://oauth2-proxy.localhost/app1"
redirect_url="http://oauth2-proxy.oauth2-proxy.localhost/oauth2/callback"
cookie_domains=".oauth2-proxy.localhost" # Required so cookie can be read on all subdomains.
whitelist_domains=".oauth2-proxy.localhost" # Required to allow redirection back to original requested target.
```

**Comment by NickMeves** [comment]
Nice! Good work getting almost there in nginx side.

If you have nginx at your fingertips and you need the Access Token in an `Authorization: Bearer` header, could you use the Lua module to programmatically make that header translation/copy?

**Comment by salmanisd** [comment]
Basically I have to use this implementation of an embedded device where we dont have Nginx Lua installed. Adding nginx lua to the firmware would be additional work + I am not sure how that can solve this issue.

I know that oauth2-proxy can pass access and id tokens (--pass-access-token and --pass-authorization-header options) .
I have also tried using $upstream_http_x_auth_request_access_token like
``` 
location /app1 {
#     auth_request /internal-auth/oauth2/auth;
      auth_request /gg;
      auth_request_set $access_token  $upstream_http_x_auth_request_access_token;
```

and then
```
location = /_oauth2_send_request {
                internal;
                proxy_method      POST;
                proxy_set_header  content-type "application/json";
                proxy_set_header  Authorization "Bearer $access_token";
                proxy_pass        http://localhost:4444/userinfo;
    }
````
but accessing localhost/app1 always returns ```401 Authorization Required```.
I must be missing something but there must be a simple way where we can just pluck the access_token sent to nginx from oauth2-proxy and use it ?

**Comment by salmanisd** [comment]
So I solved one issue.
I got access to the forwarded access token by setting ```set_xauthrequest="true"```  and then accessing it in nginx using ```$upstream_http_x_auth_request_access_token```

Previously I had just ```pass_access_token="true"``` but I couldnt access the token.

Only problem right now is that I can access the token only in location / (it uses auth_request /internal-auth/oauth2/auth;).
location /app1 (where I need this token) uses my own auth_request command and I using the ```$upstream_http_x_auth_request_access_token``` variable doesnt work there. So once again I'm just out of reach of my access control solution.

I'll probably give nginx conf some more time before finding someother application. Just wanted to document stuff here in case someone else comes looking in the future :)
